service/history/queueFactoryBase.go

```diff
@@ -122,22 +125,37 @@ func (f *QueueFactoryBase) Stop() {
 
 func NewQueueHostRateLimiter(
 	hostRPS dynamicconfig.IntPropertyFn,
-	fallBackRPS dynamicconfig.IntPropertyFn,
+	serviceResolver membership.ServiceResolver,
+	numberOfShards int32,
+	readerStuckCriticalAttempts dynamicconfig.IntPropertyFn,
 ) quotas.RateLimiter {
 	return quotas.NewDefaultOutgoingRateLimiter(
-		NewHostRateLimiterRateFn(hostRPS, fallBackRPS),
+		NewHostRateLimiterRateFn(hostRPS, serviceResolver, numberOfShards, readerStuckCriticalAttempts),
 	)
 }
 
 func NewHostRateLimiterRateFn(
 	hostRPS dynamicconfig.IntPropertyFn,
-	fallBackRPS dynamicconfig.IntPropertyFn,
+	serviceResolver membership.ServiceResolver,
+	numberOfShards int32,
+	readerStuckCriticalAttempts dynamicconfig.IntPropertyFn,
 ) quotas.RateFn {
 	return func() float64 {
 		if maxPollHostRps := hostRPS(); maxPollHostRps > 0 {
 			return float64(maxPollHostRps)
 		}
 
-		return float64(fallBackRPS())
+		shardsPerHost := int(numberOfShards) / (numHistoryHosts(serviceResolver))
+		return float64(readerStuckCriticalAttempts() * shardsPerHost)
```

Comment:
hmm, i think we should adjust based on actual shard count assigned to this host.

Reply:
Synced offline. There's a fx cycle dependency problem if we want to get the # of shards from shard controller.

Also upon host restart, service resolver doesn't have info on how many hosts are in the ring. So the rps limit may be too large. Instead, temporally set a upper bound based on the total persistence rps limit.

In long term, the dynamic rate limiter should be refactor so that RateBurst can be updated as needed instead of waiting on a 1 min timer. (see issue [here](https://github.com/temporalio/temporal/issues/3149)). With that, we can listen on membership change on service resolve and update the ratelimiter when the # of hosts changes.